utilities/transactions/pessimistic_transaction_db.h

```diff
@@ -161,6 +164,86 @@ class WritePreparedTxnDB : public PessimisticTransactionDB {
   Transaction* BeginTransaction(const WriteOptions& write_options,
                                 const TransactionOptions& txn_options,
                                 Transaction* old_txn) override;
+
+  // Check whether the transaction that wrote the value with seqeunce number seq
+  // is visible to the snapshot with sequence number snapshot_seq
+  bool IsInSnapshot(uint64_t seq, uint64_t snapshot_seq);
+  // Add the trasnaction with prepare sequence seq to the prepared list
+  void AddPrepared(uint64_t seq);
+  // Add the transaction with prepare sequence prepare_seq and commit sequence
+  // commit_seq to the commit map
+  void AddCommitted(uint64_t prepare_seq, uint64_t commit_seq);
+
+ private:
+  // A heap with the amortized O(1) complexity for erase. It uses one extra heap
+  // to keep track of erased entries that are not yet on top of the main heap.
+  class PreparedHeap {
+    std::priority_queue<uint64_t> prepared_txns_;
+    std::priority_queue<uint64_t> erased_prepared_txns_;
+
+   public:
+    bool empty() { return prepared_txns_.empty(); }
+    uint64_t top() { return prepared_txns_.top(); }
+    void push(uint64_t v) { prepared_txns_.push(v); }
+    void pop() {
+      prepared_txns_.pop();
+      while (!prepared_txns_.empty() && !erased_prepared_txns_.empty() &&
+             prepared_txns_.top() == erased_prepared_txns_.top()) {
+        prepared_txns_.pop();
+        erased_prepared_txns_.pop();
+      }
+    }
+    void erase(uint64_t seq) {
+      if (!prepared_txns_.empty() && prepared_txns_.top() == seq) {
+        prepared_txns_.pop();
+      } else {
+        erased_prepared_txns_.push(seq);
+      }
+    }
+  };
+
+  // Get the commit entry with index indexed_seq from the commit table. It
+  // returns true if such entry exists.
+  bool GetCommitEntry(uint64_t indexed_seq, uint64_t* prep_seq,
+                      uint64_t* commit_seq);
+  // Rewrite the entry with the index indexed_seq in the commit table with the
+  // commit entry <prep_seq, commit_seq>
+  void AddCommitEntry(uint64_t indexed_seq, uint64_t prep_seq,
+                      uint64_t commit_seq);
+
+  // The list of live snapshots at the last time that max_evicted_seq_ advanced.
+  // Thread-safety is provided with snapshots_mutex_.
+  std::vector<SequenceNumber> snapshots_;
+  // A heap of prepared transactions. Thread-safety is provided with
+  // prepared_mutex_.
+  PreparedHeap prepared_txns_;
+  static const uint64_t COMMIT_CACHE_SIZE = 1 << 21;  // 10m entry, 80MB size
+  struct CommitEntry {
+    uint64_t prep_seq;
```

Comment:
The thing is that later we would need to play around with bits of sequence number in order to encode two of them into 64 bits of each entry in the lock-free array. Let us move forward to uint64_t and we can always do a refactoring.